### ISSUE TITLE
add constructor assertions for ClassAssert

### ISSUE THREAD

**Issue by xenoterracide** [joel-costigliola/assertj-core#1693]
#### Example

```java
assertThat(MyClass.class).hasPackageOrBetterNoArgConstructor(); // common for frameworks
assertThat(MyClass.class).hasPublicConstructor(Integer.class, String.class); // common for
``` 
I'm not fussy on names

**Comment by xenoterracide** [comment]
that's probably fine, hmm could have `hasBeanConstructor`, but yeah if you prefer splitting it I'm ok with that.

**Comment by wettera** [comment]
Does this issue mean that we need to test whether a class have a private or protected or public constructor? I want to fix this issue
For example:
`assertThat(MyClass.class).hasPrivateConstructor`
`assertThat(MyClass.class).hasProtectedConstructor`
`assertThat(MyClass.class).hasPublicConstructor`
@xenoterracide @joel-costigliola

**Comment by wettera** [comment]
sorry......
forget the brackets and semicolon after each asseration.

**Comment by scordio** [comment]
Thinking out loud, would it make sense to introduce assertions for `java.lang.reflect.Constructor`?

And `AbstractClassAssert` would have a new `getConstructor(Class<?>...)` method.

Something like:
```java
assertThat(MyClass.class).getConstructor()  // default constructor
                         .isPublic();
assertThat(MyClass.class).getConstructor(Integer.class, String.class)
                         .isPackagePrivate();
```

**Comment by wettera** [comment]
But one class could have multiple constructors, these constructors may be private, protected, or public.

**Comment by wettera** [comment]
`.getConstructor()` may get a constructor list. Maybe in the list one constructor is parvate, another contructor is public.

**Comment by scordio** [comment]
> But one class could have multiple constructors, these constructors may be private, protected, or public.

The `getConstructor(Class<?>...)` I proposed would have `Class#getDeclaredConstructor(Class<?>...)` behind, where the resulting constructor depends on the parameter types given as input, so only one constructor at a time would be returned.

**Comment by wettera** [comment]
> > But one class could have multiple constructors, these constructors may be private, protected, or public.
> 
> The `getConstructor(Class<?>...)` I proposed would have `Class#getDeclaredConstructor(Class<?>...)` behind, where the resulting constructor depends on the parameter types given as input, so only one constructor at a time would be returned.
> 
> This is actually a second step. It would be enough to have `ConstructorAssert` which provides assertions on the modifiers. Something like this could be written:
> 
> ```java
> assertThat(MyClass.class.getDeclaredConstructor()).isPublic();
> ```

Both `MyClass.class.getDeclaredConstructors()` and `MyClass.class.getConstructors()`  parameter list is empty. I checked just now.

**Comment by wettera** [comment]
> > But one class could have multiple constructors, these constructors may be private, protected, or public.
> 
> The `getConstructor(Class<?>...)` I proposed would have `Class#getDeclaredConstructor(Class<?>...)` behind, where the resulting constructor depends on the parameter types given as input, so only one constructor at a time would be returned.
> 
> This is actually a second step. It would be enough to have `ConstructorAssert` which provides assertions on the modifiers. Something like this could be written:
> 
> ```java
> assertThat(MyClass.class.getDeclaredConstructor()).isPublic();
> ```

Sorry I make a mistake, `instance.getClass().getDeclaredConstructors()` and `instance.getClass().getConstructors()` parameter list is empty.

**Comment by wettera** [comment]
> > But one class could have multiple constructors, these constructors may be private, protected, or public.
> 
> The `getConstructor(Class<?>...)` I proposed would have `Class#getDeclaredConstructor(Class<?>...)` behind, where the resulting constructor depends on the parameter types given as input, so only one constructor at a time would be returned.
> 
> This is actually a second step. It would be enough to have `ConstructorAssert` which provides assertions on the modifiers. Something like this could be written:
> 
> ```java
> assertThat(MyClass.class.getDeclaredConstructor()).isPublic();
> ```
Sorry.....
 `MyClass.class.getDeclaredConstructors()`
 `MyClass.class.getConstructors()` 
`instance.getClass().getDeclaredConstructors()`
`instance.getClass().getDeclaredConstructors()`
can have parameter type as input and return one unique constructor. 
I'm sorry I was so careless.

**Comment by scordio** [comment]
@xenoterracide @wettera we discussed the topic internally and we would prefer a separate `ConstructorAssert` to avoid bloating the `ClassAssert` API. Do you see any issue with it?

**Comment by xenoterracide** [comment]
nope, though you might consider allowing chaining from the class assert, that way it's easily discoverable, but I'll leave that up to you. Since my biggest priority was to ensure that I had a public noargs constructor.

**Comment by scordio** [comment]
> you might consider allowing chaining from the class assert

Yes, chaining is definitely in the roadmap

**Comment by JitendraProgrammer** [comment]
Hi Guys,

Can you guide me how can i contribute here
